### PR TITLE
[Feature] Support HorNet Backbone

### PR DESCRIPTION
**Pull request by okotaku**
## Motivation

[HorNet: Efficient High-Order Spatial Interactions with Recursive Gated Convolutions](https://arxiv.org/pdf/2207.14284v2.pdf)
[github](https://github.com/raoyongming/HorNet)

## Weights

[hornet-tiny](https://tmp.link/f/63158733df58b)
[hornet-tiny-gf](https://tmp.link/f/6315876bbed10)
[hornet-small](https://tmp.link/f/63158901bf8e9)
[hornet-small-gf](https://tmp.link/f/6315891a51464)
[hornet-base](https://tmp.link/f/6315878fd7712)
[hornet-base-gf](https://tmp.link/f/631587b45f460)
[hornet-large](https://tmp.link/f/63158801e8ec8)
[hornet-large-gf](https://tmp.link/f/631588e615904)
[hornet-large-gf384](https://tmp.link/f/63158850ea0f8)

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by Ezra-Yu** [comment]
Great!  Thanks for your contribution!

File: mmcls/models/backbones/hornet.py [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+    """
+    arch_zoo = {
+        **dict.fromkeys(['t', 'tiny'],
+                        {'base_dim': 64,
+                         'depths': [2, 3, 18, 2],
+                         'orders': [2, 3, 4, 5],
+                         'gflayers': ['DWConv', 'DWConv', 'DWConv', 'DWConv'],
+                         'hs': [14, 14, 14, 14],
+                         'ws': [8, 8, 8, 8],
+                         's': 1 / 3}),
```
Comment (Ezra-Yu) [inline mmcls/models/backbones/hornet.py:305]: I think this value can move into the backbone class  as `scale`

File: mmcls/models/backbones/hornet.py [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+            assert isinstance(arch, dict) and set(arch) == essential_keys, \
+                f'Custom arch needs a dict with keys {essential_keys}'
+            self.arch_settings = arch
+
+        self.out_indices = out_indices
+        self.frozen_stages = frozen_stages
+        self.with_cp = with_cp
+        self.gap_before_final_norm = gap_before_final_norm
+
+        dims = [
```
Comment (Ezra-Yu) [inline mmcls/models/backbones/hornet.py:419]:
```
base_dim = self.arch_settings['base_dim']
dims = list(map(lambda x : 2**x * base_dim, range(4)))
```

File: mmcls/models/backbones/hornet.py [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+        for i in range(3):
+            downsample_layer = nn.Sequential(
+                HorNetLayerNorm(
+                    dims[i], eps=1e-6, data_format='channels_first'),
+                nn.Conv2d(dims[i], dims[i + 1], kernel_size=2, stride=2),
+            )
+            self.downsample_layers.append(downsample_layer)
+
+        self.stages = nn.ModuleList()
+        dp_rates = [
```
Comment (Ezra-Yu) [inline mmcls/models/backbones/hornet.py:439]: use `dpr` for consistency with other implementations.

**Review by Ezra-Yu — CHANGES_REQUESTED** [review]
1. please add the model in [`doc/apis/models.rst`](https://github.com/open-mmlab/mmclassification/blame/master/docs/en/api/models.rst) #backbone so that you can view your model docstring in the doc by clicking the link in the PR page.
image
2.  you can add convert script in `tools/convert_models/` folder since there are many checkpoints.

File: configs/hornet/metafile.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        - Dataset: ImageNet-1k
+          Metrics:
+            Top 1 Accuracy: 83.98
+            Top 5 Accuracy: 96.77
+          Task: Image Classification
+    Weights:
+    Config: configs/hornet/hornet-small-gf_8xb128_in1k.py
+  - Name: hornet-base_8xb128_in1k
+    Metadata:
+      FLOPs: 87260000            # 87.26M
```
[the remark this answers is not in the document]
Comment (okotaku) [reply configs/hornet/metafile.yml:73]: VAN metafile has same problem. I will fix both.

File: mmcls/models/backbones/hornet.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+        return x
+
+
+class gnConv(nn.Module):
+    """A gnConv of HorNet.
+
+    Args:
+        dim (int): Number of input channels.
+        order (int): Order of gnConv.
+            Default to 5
```
[the remark this answers is not in the document]
Comment (okotaku) [reply mmcls/models/backbones/hornet.py]: Sounds good.

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

**Pull request merged**